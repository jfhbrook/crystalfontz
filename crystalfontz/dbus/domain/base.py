from typing import (
    ClassVar,
    List,
    Optional,
    Protocol,
    Self,
    Type,
    Union,
)


class TypeProtocol(Protocol):
    t: ClassVar[str]


def t(*args: Union[str, Type[TypeProtocol]]) -> str:
    type_ = ""

    for arg in args:
        if isinstance(arg, str):
            type_ += arg
        else:
            type_ += arg.t

    return type_


def struct(*args: Union[str, Type[TypeProtocol]]) -> str:
    return t("(", *args, ")")


def array(of: Union[str, Type[TypeProtocol]]) -> str:
    return f"a{t(of)}"


class NoneM:
    t: ClassVar[str] = ""


OptUintT = int


class OptUintM:
    """
    Map optional positive integers to and from dbus types.

    This class treats values less than 0 (ie, -1) as representing None.
    """

    t: ClassVar[str] = "i"
    none: ClassVar[OptUintT] = -1

    @staticmethod
    def unpack(r: OptUintT) -> Optional[int]:
        return r if r >= 0 else None

    @classmethod
    def pack(cls: Type[Self], r: Optional[int]) -> OptUintT:
        return r if r is not None else cls.none


OptPosFloatT = float


class OptPosFloatM:
    """
    Map optional positive floats to and from dbus types.

    This class treats values less than 0 (ie, -1.0) as representing None.
    """

    t: ClassVar[str] = "d"
    none: ClassVar[OptPosFloatT] = -1.0

    @staticmethod
    def unpack(t: OptPosFloatT) -> Optional[float]:
        return t if t >= 0 else None

    @classmethod
    def pack(cls: Type[Self], t: Optional[float]) -> OptPosFloatT:
        return t if t is not None else cls.none


Uint16T = int


class Uint16M:
    t: ClassVar[str] = "q"


AddressT = Uint16T


class AddressM(Uint16M):
    t: ClassVar[str] = Uint16M.t


ByteT = int


class ByteM:
    t: ClassVar[str] = "y"


IndexT = ByteT


class IndexM(ByteT):
    t: ClassVar[str] = ByteM.t


PositionT = Uint16T


class PositionM(ByteT):
    t: ClassVar[str] = ByteM.t


BytesT = List[int]


class BytesM:
    """
    Map bytes to and from dbus types.
    """

    t: ClassVar[str] = array(ByteM)

    @staticmethod
    def unpack(buff: BytesT) -> bytes:
        return bytes(buff)

    # TODO: This may not be what the dbus client expects...
    @staticmethod
    def pack(buff: bytes) -> BytesT:
        return list(buff)


OptBytesT = BytesT


class OptBytesM:
    """
    Map optional bytes to and from dbus types.
    """

    t: ClassVar[str] = BytesM.t
    none: ClassVar[BytesT] = list()

    @staticmethod
    def unpack(buff: BytesT) -> Optional[bytes]:
        if not buff:
            return None
        return BytesM.unpack(buff)

    @classmethod
    def pack(cls: Type[Self], buff: Optional[bytes]) -> BytesT:
        if buff is None:
            return cls.none
        return BytesM.pack(buff)


ModelT = str


class ModelM:
    t: ClassVar[str] = "s"


RevisionT = str


class RevisionM:
    t: ClassVar[str] = "s"
    none: ClassVar[str] = ""

    @classmethod
    def pack(cls: Type[Self], revision: Optional[str]) -> str:
        return revision or cls.none

    @classmethod
    def unpack(cls: Type[Self], revision: str) -> Optional[str]:
        return revision if revision != cls.none else None


TimeoutT = float


class TimeoutM(OptPosFloatM):
    """
    Map timeout parameters to and from dbus types.

    TimeoutM is an alias for OptFloatM.
    """

    pass


RetryTimesT = int


class RetryTimesM(OptUintM):
    """
    Map retry times parameters to and from dbus types.

    RetryTimesM is an alias for OptFloatM.
    """

    pass


class OkM:
    t: ClassVar[str] = "b"
