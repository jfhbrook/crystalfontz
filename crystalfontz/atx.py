from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Optional, Self

AUTO_POLARITY = 0x01


class AtxPowerSwitchFunction(Enum):
    RESET_INVERT = 0x02
    POWER_INVERT = 0x04
    LCD_OFF_IF_HOST_IS_OFF = 0x10
    KEYPAD_RESET = 0x20
    KEYPAD_POWER_ON = 0x40
    KEYPAD_POWER_OFF = 0x80


@dataclass
class AtxPowerSwitchFunctionalitySettings:
    functions: Iterable[AtxPowerSwitchFunction]
    auto_polarity: bool = True
    power_pulse_length_seconds: Optional[float] = None

    def to_bytes(self: Self) -> bytes:
        functions: int = 0
        for function in self.functions:
            functions ^= function.value
        if self.auto_polarity:
            functions ^= AUTO_POLARITY

        packed: bytes = functions.to_bytes()

        if self.power_pulse_length_seconds is not None:
            pulse_length = int(self.power_pulse_length_seconds * 32)

            if pulse_length < 1:
                raise ValueError(f"Pulse length can not be less than {1/32}")

            packed += min(pulse_length, 255).to_bytes()

        return packed
