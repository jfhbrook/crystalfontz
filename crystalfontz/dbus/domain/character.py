from typing import ClassVar

from crystalfontz.character import SpecialCharacter

SpecialCharacterT = int


class SpecialCharacterM:
    t: ClassVar[str] = "t"

    @staticmethod
    def unpack(character: SpecialCharacterT) -> SpecialCharacter:
        raise NotImplementedError("load")
