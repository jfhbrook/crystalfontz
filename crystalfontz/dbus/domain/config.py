from typing import Any, cast, ClassVar, Optional, Self, Tuple, Type

from crystalfontz.config import Config
from crystalfontz.dbus.domain.base import (
    ModelM,
    ModelT,
    RetryTimesM,
    RetryTimesT,
    RevisionM,
    RevisionT,
    struct,
    TimeoutM,
    TimeoutT,
)
from crystalfontz.dbus.domain.baud import BaudRateM, BaudRateT

FileT = str


class FileM:
    t: ClassVar[str] = "s"
    none: ClassVar[str] = ""

    @classmethod
    def pack(cls: Type[Self], file: Optional[str]) -> str:
        return file or cls.none


PortT = str


class PortM:
    t: ClassVar[str] = "s"


ConfigT = Tuple[
    FileT, PortT, ModelT, RevisionT, RevisionT, BaudRateT, TimeoutT, RetryTimesT
]


class ConfigM:
    """
    Map a config to and from dbus types.
    """

    t: ClassVar[str] = struct(
        FileM,
        PortM,
        ModelM,
        RevisionM,
        RevisionM,
        BaudRateM,
        TimeoutM,
        RetryTimesM,
    )

    @staticmethod
    def pack(config: Config) -> ConfigT:
        return (
            FileM.pack(config.file),
            config.port,
            config.model,
            RevisionM.pack(config.hardware_rev),
            RevisionM.pack(config.firmware_rev),
            config.baud_rate,
            TimeoutM.pack(config.timeout),
            RetryTimesM.pack(config.retry_times),
        )

    @staticmethod
    def unpack(config: ConfigT) -> Config:
        (
            file,
            port,
            model,
            hardware_rev,
            firmware_rev,
            baud_rate,
            timeout,
            retry_times,
        ) = config

        return cast(Any, Config)(
            file=file,
            port=port,
            model=model,
            hardware_rev=RevisionM.unpack(hardware_rev),
            firmware_rev=RevisionM.unpack(firmware_rev),
            baud_rate=baud_rate,
            timeout=TimeoutM.unpack(timeout),
            retry_times=RetryTimesM.unpack(retry_times),
        )
