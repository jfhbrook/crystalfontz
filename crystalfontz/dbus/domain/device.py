from typing import ClassVar, Tuple

from crystalfontz.dbus.domain.base import (
    BytesM,
    BytesT,
    ModelM,
    ModelT,
    RevisionM,
    RevisionT,
    struct,
    t,
    Uint16M,
    Uint16T,
)
from crystalfontz.device import Device, DeviceStatus, lookup_device

WidthT = Uint16T


class WidthM(Uint16M):
    pass


HeightT = Uint16T


class HeightM(Uint16M):
    pass


DeviceT = Tuple[ModelT, RevisionT, RevisionT, WidthT, HeightT, WidthT, HeightT, Uint16T]


class DeviceM:
    t: ClassVar[str] = struct(
        ModelM, RevisionM, RevisionM, WidthM, HeightM, WidthM, HeightM, Uint16M
    )

    @staticmethod
    def pack(device: Device) -> DeviceT:
        return (
            device.model,
            device.hardware_rev,
            device.firmware_rev,
            device.lines,
            device.columns,
            device.character_width,
            device.character_height,
            device.n_temperature_sensors,
        )

    @staticmethod
    def unpack(device: DeviceT) -> Device:
        model = device[0]
        hardware_rev = device[1]
        firmware_rev = device[2]

        return lookup_device(model, hardware_rev, firmware_rev)


# Includes enough device info to unpack the status
DeviceStatusT = Tuple[ModelT, RevisionT, RevisionT, BytesT]


class DeviceStatusM:
    t: ClassVar[str] = t(ModelM, RevisionM, RevisionM, BytesM)

    @staticmethod
    def pack(_status: DeviceStatus) -> DeviceStatusT:
        raise NotImplementedError("pack")

    @staticmethod
    def unpack(_status: DeviceStatusT) -> DeviceStatus:
        raise NotImplementedError("unpack")
