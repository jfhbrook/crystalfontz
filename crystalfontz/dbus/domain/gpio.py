from typing import ClassVar, Dict, Optional, Self, Tuple, Type

from crystalfontz.dbus.domain.base import ByteM, ByteT
from crystalfontz.gpio import GpioDriveMode, GpioFunction, GpioSettings, GpioState

GpioStateT = Tuple[bool, bool, bool]


class GpioStateM:
    t: ClassVar[str] = "bbb"

    @staticmethod
    def pack(state: GpioState) -> GpioStateT:
        return (state.state, state.falling, state.rising)

    @staticmethod
    def unpack(state: GpioStateT) -> GpioState:
        st, falling, rising = state
        return GpioState(st, falling, rising)


GpioFunctionT = ByteT

GPIO_FUNCTIONS: Dict[ByteT, GpioFunction] = {
    function.value: function for function in GpioFunction
}


class GpioFunctionM:
    """
    Map GpioFunction to and from dbus types.
    """

    t: ClassVar[str] = ByteM.t

    @staticmethod
    def pack(function: GpioFunction) -> GpioFunctionT:
        return function.value

    @staticmethod
    def unpack(function: GpioFunctionT) -> GpioFunction:
        return GPIO_FUNCTIONS[function]


GpioDriveModeT = ByteT

GPIO_DRIVE_MODES: Dict[ByteT, GpioDriveMode] = {
    drive_mode.value: drive_mode for drive_mode in GpioDriveMode
}


class GpioDriveModeM:
    """
    Map GpioDriveMode to and from dbus types.
    """

    t: ClassVar[str] = ByteM.t

    @staticmethod
    def pack(drive_mode: GpioDriveMode) -> GpioDriveModeT:
        return drive_mode.value

    @staticmethod
    def unpack(drive_mode: GpioDriveModeT) -> GpioDriveMode:
        return GPIO_DRIVE_MODES[drive_mode]


GpioSettingsT = ByteT


class GpioSettingsM:
    t: ClassVar[str] = ByteM.t

    @staticmethod
    def pack(settings: GpioSettings) -> GpioSettingsT:
        return settings.to_bytes()[0]

    @staticmethod
    def unpack(settings: GpioSettingsT) -> GpioSettings:
        return GpioSettings.from_byte(settings)


OptGpioSettingsT = int


class OptGpioSettingsM:
    t: ClassVar[str] = "n"
    none: ClassVar[int] = -1

    @classmethod
    def pack(cls: Type[Self], settings: Optional[GpioSettings]) -> OptGpioSettingsT:
        return GpioSettingsM.pack(settings) if settings is not None else cls.none

    @staticmethod
    def unpack(settings: OptGpioSettingsT) -> Optional[GpioSettings]:
        return GpioSettingsM.unpack(settings) if settings >= 0 else None
