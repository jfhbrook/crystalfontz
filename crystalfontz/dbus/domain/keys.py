from typing import ClassVar, Set, Tuple

from crystalfontz.dbus.domain.base import ByteM, OptFloatM, OptFloatT, struct
from crystalfontz.keys import (
    KeyPress,
    KeyState,
    KeyStates,
    KP_DOWN,
    KP_ENTER,
    KP_EXIT,
    KP_LEFT,
    KP_RIGHT,
    KP_UP,
)

KeyPressT = int

KEYPRESSES: Set[KeyPress] = {KP_UP, KP_ENTER, KP_EXIT, KP_LEFT, KP_RIGHT, KP_DOWN}


class KeyPressM(ByteM):
    t: ClassVar[str] = ByteM.t

    @staticmethod
    def unpack(keypress: int) -> KeyPress:
        if keypress not in KEYPRESSES:
            raise ValueError(f"{keypress} is not a valid key press")
        return keypress


KeyStateT = Tuple[bool, bool, bool]
KeyStatesT = Tuple[
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
]


class KeyStateM:
    t: ClassVar[str] = struct("bbb")

    @staticmethod
    def pack(state: KeyState) -> KeyStateT:
        return (state.pressed, state.pressed_since, state.released_since)

    @staticmethod
    def unpack(state: KeyStateT) -> KeyState:
        pressed, pressed_since, released_since = state
        return KeyState(
            pressed=pressed, pressed_since=pressed_since, released_since=released_since
        )


class KeyStatesM:
    t: ClassVar[str] = struct(KeyStateM.t) * 6

    @staticmethod
    def pack(states: KeyStates) -> KeyStatesT:
        return (
            KeyStateM.pack(states.up),
            KeyStateM.pack(states.enter),
            KeyStateM.pack(states.exit),
            KeyStateM.pack(states.left),
            KeyStateM.pack(states.right),
            KeyStateM.pack(states.down),
        )

    @staticmethod
    def unpack(states: KeyStatesT) -> KeyStates:
        up, enter, exit, left, right, down = states
        return KeyStates(
            up=KeyStateM.unpack(up),
            enter=KeyStateM.unpack(enter),
            exit=KeyStateM.unpack(exit),
            left=KeyStateM.unpack(left),
            right=KeyStateM.unpack(right),
            down=KeyStateM.unpack(down),
        )


KeypadBrightnessT = OptFloatT


class KeypadBrightnessM(OptFloatM):
    """
    Map `Optional[float]` to and from `KeypadBrightnessT` (`float`).

    `KeypadBrightnessM` is an alias for `OptFloatM`.
    """

    t: ClassVar[str] = OptFloatM.t
