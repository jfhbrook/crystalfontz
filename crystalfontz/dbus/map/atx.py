from typing import ClassVar, List, Tuple

from crystalfontz.atx import AtxPowerSwitchFunction, AtxPowerSwitchFunctionalitySettings
from crystalfontz.dbus.map.base import array, t

AtxPowerSwitchFunctionT = str
AtxPowerSwitchFunctionalitySettingsT = Tuple[List[str], bool, float]


class AtxPowerSwitchFunctionM:
    t: ClassVar[str] = "s"

    @staticmethod
    def unpack(function: AtxPowerSwitchFunctionT) -> AtxPowerSwitchFunction:
        return AtxPowerSwitchFunction(function)


class AtxPowerSwitchFunctionalitySettingsM:
    t: ClassVar[str] = t(array(AtxPowerSwitchFunctionM), "bd")

    @staticmethod
    def unpack(
        settings: AtxPowerSwitchFunctionalitySettingsT,
    ) -> AtxPowerSwitchFunctionalitySettings:
        functions, auto_polarity, power_pulse_length = settings
        return AtxPowerSwitchFunctionalitySettings(
            functions={AtxPowerSwitchFunctionM.unpack(name) for name in functions},
            auto_polarity=auto_polarity,
            power_pulse_length_seconds=power_pulse_length,
        )
