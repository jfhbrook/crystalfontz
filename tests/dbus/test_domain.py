from typing import Any, Callable, Optional

import pytest

from crystalfontz.dbus.domain.atx import (
    AtxPowerSwitchFunction,
    AtxPowerSwitchFunctionalitySettingsM,
)

ValidateFn = Callable[[Any, Any], None]


def validate_eq(actual: Any, expected: Any) -> None:
    assert actual == expected


def validate_is(actual: Any, expected: Any) -> None:
    assert isinstance(actual, expected.__class__)


@pytest.mark.skip
def test_domain_pack_unpack(
    entity: Any, map_cls: Any, validate: Optional[ValidateFn], snapshot
) -> None:
    packed = map_cls.pack(entity)

    assert packed == snapshot

    if hasattr(map_cls, "unpack"):
        unpacked = map_cls.pack(packed)
        if validate:
            validate(unpacked, entity)
        else:
            assert unpacked == entity


@pytest.mark.parametrize(
    "packed,map_cls",
    [
        (
            ([AtxPowerSwitchFunction.KEYPAD_RESET.value], False, True, True, 1.0),
            AtxPowerSwitchFunctionalitySettingsM,
        )
    ],
)
def test_domain_unpack_pack(packed: Any, map_cls: Any, snapshot) -> None:
    entity = map_cls.unpack(packed)

    assert entity == snapshot

    if hasattr(map_cls, "pack"):
        repacked = map_cls.pack(entity)
        assert repacked == packed
