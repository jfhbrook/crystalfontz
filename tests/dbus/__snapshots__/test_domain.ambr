# serializer version: 1
# name: test_domain_unpack_pack[packed0-AtxPowerSwitchFunctionalitySettingsM]
  Functions enabled: KEYPAD_RESET
  Auto-Polarity Enabled: no
  Reset Inverted: yes
  Power Inverted: yes
  Power Pulse Length (seconds): 1.0
# ---
